File: Sujin/Ch12_Graph/332_Reconstruct_ltinerary.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <queue>
using namespace std;

struct compare{
    bool operator()(string& a, string& b){
        return a > b; // 사전순 정렬
    }
};

class Solution {
public:
    map<string, priority_queue<string, vector<string>, compare>> m;
    vector<string> res; // 결과를 반환할 벡터 res
    
    void DFS(string s){
        while(!m[s].empty()){
            string next = m[s].top();
            m[s].pop();
            DFS(next);
        }
        res.insert(res.begin(), s);
        cout<<s<<endl;
    }
    
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        
        for(int i=0; i<tickets.size(); i++){
            m[tickets[i][0]].push(tickets[i][1]);
        }
        DFS("JFK");
        
        return res;
    }
};

File: Suyeon/ch12_Graph/332_Reconstruct_Ltinerary.py
import collections

class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        info = collections.defaultdict(list)

        for ticket in sorted(tickets, reverse=True):
            info[ticket[0]].append(ticket[1])

        for list in info:
            list.sort(reverse=True)

        result = []

        def dfs(start):

            while info[start]:
                dfs(info[start].pop())

            result.append(start)

        dfs('JFK')

        return result



